### DISCUSSION TITLE
No location update when vehicle is moving

### DISCUSSION THREAD

**Discussion by sshaikh** [bimmerconnected/bimmer_connected#631] [Q&A] [answered]
The behaviour in a vehicle we are connected to only updates the location when for eg the door lock status changes. I understand that this is by design for some model of vehicle. Is there a way to confirm this (ie might providing signatures or some such improve the situation for this model?). Generally is there a registry of features against models that are confirmed to work/behave?

This makes location a bit unreliable even as a rough guide. Does anyone have any workaround to this? We understood that `find vehicle` prompts the vehicle to report in but this didn't seem to do anything in this instance (we still got the update on door lock).

Given that locking the doors does update the location, would (in theory) cycling the locks be a way to force update the location? Alternatively is there a more benign command that would update the location as a side effect?

**Comment by rikroe** [comment]
Are you talking about the Home Assistant integration or the CLI?

Generally to my knowledge, vehicles report their current state when they are shut off and locked, i.e. the vehicle's ICE (if available) is turned off and the car locked.
Then, the current state (including location) is sent to BMW servers and can be fetched by us.

Triggering a service (like locking/unlocking) always forces a refresh of the data from BMW servers. 
If this is not the case for the find vehicle service, there might be a bug in the library.

**Reply by sshaikh** [reply]
This is all via HA.

We're trying to write an automation to alert if the car doors have been left unlocked while the car is stationary. If location won't be sent until the car is locked, then that's a bit of a non-starter :).

Our understanding of the find vehicle service was to request the vehicle update its location on the BMW servers. If not that, what does it do?

**Reply by rikroe** [reply]
Some time ago there was a discussion that some cars do support location updates while driving.

Can you check with a second person using the MyBMW app if it is updating automatically or when using the find vehicle service inside the app **while the car is moving**?

**Reply by sshaikh** [reply]
Triggering `vehiclefinder` does seem to at least update the timestamp on the location in the status. I will experiment the next time the car is in motion.

**Reply by sshaikh** [reply]
Here is the output when I invoked `vehiclefinder` while the car was moving, although it seems to be happening even now while stationary:

```
❯ bimmerconnected vehiclefinder $USERNAME $PASSWORD $REGION $VIN
MyBMWAPIError due to HTTPStatusError: Bad Request -
Traceback (most recent call last):
  File "/home/projects/bimmer_connected/lib/python3.12/site-packages/bimmer_connected/api/client.py", line 71, in raise_for_status_event_handler
    response.raise_for_status()
  File "/home/projects/bimmer_connected/lib/python3.12/site-packages/httpx/_models.py", line 761, in raise_for_status
    raise HTTPStatusError(message, request=request, response=self)
httpx.HTTPStatusError: Client error '400 Bad Request' for url 'https://cocoapi.bmwgroup.com/eadrax-vrccs/v3/presentation/remote-commands/eventPosition?eventId=xxx'
For more information check: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/projects/bimmer_connected/bin/bimmerconnected", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/projects/bimmer_connected/lib/python3.12/site-packages/bimmer_connected/cli.py", line 341, in main
    loop.run_until_complete(args.func(args))
  File "/usr/lib/python3.12/asyncio/base_events.py", line 687, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/home/projects/bimmer_connected/lib/python3.12/site-packages/bimmer_connected/cli.py", line 208, in vehicle_finder
    status = await vehicle.remote_services.trigger_remote_vehicle_finder()
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/projects/bimmer_connected/lib/python3.12/site-packages/bimmer_connected/vehicle/remote_services.py", line 313, in trigger_remote_vehicle_finder
    result = await self._get_event_position(status.event_id)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/projects/bimmer_connected/lib/python3.12/site-packages/bimmer_connected/vehicle/remote_services.py", line 327, in _get_event_position
    response = await client.post(
               ^^^^^^^^^^^^^^^^^^
  File "/home/projects/bimmer_connected/lib/python3.12/site-packages/httpx/_client.py", line 1892, in post
    return await self.request(
           ^^^^^^^^^^^^^^^^^^^
  File "/home/projects/bimmer_connected/lib/python3.12/site-packages/httpx/_client.py", line 1574, in request
    return await self.send(request, auth=auth, follow_redirects=follow_redirects)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/projects/bimmer_connected/lib/python3.12/site-packages/httpx/_client.py", line 1661, in send
    response = await self._send_handling_auth(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/projects/bimmer_connected/lib/python3.12/site-packages/httpx/_client.py", line 1689, in _send_handling_auth
    response = await self._send_handling_redirects(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/projects/bimmer_connected/lib/python3.12/site-packages/httpx/_client.py", line 1747, in _send_handling_redirects
    raise exc
  File "/home/projects/bimmer_connected/lib/python3.12/site-packages/httpx/_client.py", line 1729, in _send_handling_redirects
    await hook(response)
  File "/home/projects/bimmer_connected/lib/python3.12/site-packages/bimmer_connected/api/client.py", line 73, in raise_for_status_event_handler
    await handle_httpstatuserror(ex, log_handler=_LOGGER)
  File "/home/projects/bimmer_connected/lib/python3.12/site-packages/bimmer_connected/api/utils.py", line 86, in handle_httpstatuserror
    raise _ex_to_raise(_err_message) from ex
bimmer_connected.models.MyBMWAPIError: HTTPStatusError: Bad Request -
```

**Reply by sshaikh** [reply]
Seems to work again with `bimmer-connected==0.16.0.0b2` (only tested while stationary), but then I'm sure it was working previously too. Will report back after testing in motion.

**Reply by sshaikh** [reply]
I got to try the library while the vehicle is in motion. The heading seems to be live and up to date but the long and lat seem to be static. I'm also getting the following feedback from the CLI:

```
~/projects/bimmer_connected via 🐍 v3.12.4 (bimmer_connected) took 38s
❯ bimmerconnected vehiclefinder $USERNAME $PASSWORD $REGION $VIN 100 10
Error retrieving vehicle position. Unknown position: Set observer position to retrieve vehicle coordinates!
ExecutionState.EXECUTED
{'gps_position': GPSPosition(latitude=xxx, longitude=yyy), 'heading': 198}
```

Where xxx and yyy don't seem to change. I've tried various values of "observer position" which I believe to be the last two positional arguments?

**Reply by rikroe** [reply]
Yes, the two last positional arguments are lat and lon.
Currently we expect them to be floats, so try `50.0 0.0`.
Also I would be interested if the MyBMW app updates the vehicle location while it is moving.

**Reply by sshaikh** [reply]
No luck:

```
❯ bimmerconnected vehiclefinder $USERNAME $PASSWORD $REGION $VIN 100.0 0.0
Error retrieving vehicle position. Unknown position: Set observer position to retrieve vehicle coordinates!
ExecutionState.EXECUTED
{'gps_position': GPSPosition(latitude=x, longitude=y), 'heading': 25}
```

I'll have a go at debugging this.

**Reply by sshaikh** [reply]
The observer position wasn't being set in my version (`bimmer-connected==0.16.0.0b2`). After adding it, the guard fails if 0.0 is used :):
```
    if args.lat and args.lng:
        account.set_observer_position(args.lat, args.lng)
```

I've not tried calling this while the vehicle is in motion however.

**Reply by rikroe** [reply]
Ah, then try `0.1 0.1`. This should pass the test. For most vehicles, the actual position does not matter (however it is required by the API to provide it).

**Reply by sshaikh** [reply, marked as answer]
I had the chance of running this while the vehicle was in motion:

```
❯ bimmerconnected vehiclefinder $USERNAME $PASSWORD $REGION $VIN 0.1 0.1
Error retrieving vehicle position. Vehicle in use: Vehicle Finder only works when your vehicle's ignition is turned off. Try again with the ignition off.
ExecutionState.EXECUTED
{'gps_position': GPSPosition(latitude=xxx, longitude=yyy), 'heading': 201}
```

I don't think I've seen that before (it may have been suppressed because of the observer setting). Does this confirm that the service doesn't work unless the car is off? (I'm yet to confirm the behaviour with the app).

**Reply by rikroe** [reply]
As this is not an error message written by me, this is the message of the API.
So at least for your vehicle, this only seems to work when the ignition is off.

**Reply by sshaikh** [reply]
Great, thanks for the confirmation. In that case it sounds like there's no further action to take on this ticket? Can close it if so.

**Reply by rikroe** [reply]
We can keep this as answered, as it might be interesting for future users.
I'll leave the issue open to be reminded to fix the GPS location detection sometime

**Reply by sincze** [reply]
Mmm..
While using the CLI (
Every 5 min to prevent 429) with my i4 location is updated. I grab the LAT and LON and push that to TRACCAR.
It also updates the address.

**Reply by sincze** [reply]
I just noticed my car now only sends the location when it is parked. No more during moving (actual driving of the car). And based on my traccar logs I can see this happening 01-10-2024 (October 1st) before that I had a location for every query I did using bimmerconnected.
